File: netlify/functions/google-refresh-token.js
const fetch = require('node-fetch');

exports.handler = async (event, context) => {
  const headers = {
    'Content-Type': 'application/json',
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Headers': 'Content-Type',
    'Access-Control-Allow-Methods': 'POST, OPTIONS'
  };

  if (event.httpMethod === 'OPTIONS') {
    return { statusCode: 200, headers, body: '' };
  }

  try {
    const body = JSON.parse(event.body || '{}');
    const refreshToken = body.refresh_token;

    if (!refreshToken) {
      return {
        statusCode: 400,
        headers,
        body: JSON.stringify({ error: 'Missing refresh_token' })
      };
    }

    const clientId = process.env.GOOGLE_CLIENT_ID;
    const clientSecret = process.env.GOOGLE_CLIENT_SECRET;

    if (!clientId || !clientSecret) {
      return {
        statusCode: 500,
        headers,
        body: JSON.stringify({ error: 'Google OAuth not configured' })
      };
    }

    const tokenUrl = 'https://oauth2.googleapis.com/token';
    const tokenData = new URLSearchParams({
      refresh_token: refreshToken,
      client_id: clientId,
      client_secret: clientSecret,
      grant_type: 'refresh_token'
    });

    const response = await fetch(tokenUrl, {
      method: 'POST',
      body: tokenData
    });

    const tokens = await response.json();

    if (tokens.error) {
      return {
        statusCode: 400,
        headers,
        body: JSON.stringify({ error: tokens.error_description || 'Token refresh failed' })
      };
    }

    return {
      statusCode: 200,
      headers,
      body: JSON.stringify({
        access_token: tokens.access_token,
        expires_in: tokens.expires_in
      })
    };

  } catch (error) {
    return {
      statusCode: 500,
      headers,
      body: JSON.stringify({ error: error.message })
    };
  }
};

File: netlify/functions/google-refresh-token.py
import json
import os
import requests

def handler(event, context):
    """Refresh Google OAuth access token"""
    
    headers = {
        "Content-Type": "application/json",
        "Access-Control-Allow-Origin": "*",
        "Access-Control-Allow-Headers": "Content-Type",
        "Access-Control-Allow-Methods": "POST, OPTIONS"
    }
    
    if event.get("httpMethod") == "OPTIONS":
        return {"statusCode": 200, "headers": headers, "body": ""}
    
    try:
        # Get refresh token from body
        body = json.loads(event.get("body", "{}"))
        refresh_token = body.get("refresh_token")
        
        if not refresh_token:
            return {
                "statusCode": 400,
                "headers": headers,
                "body": json.dumps({"error": "Missing refresh_token"})
            }
        
        client_id = os.environ.get("GOOGLE_CLIENT_ID")
        client_secret = os.environ.get("GOOGLE_CLIENT_SECRET")
        
        if not client_id or not client_secret:
            return {
                "statusCode": 500,
                "headers": headers,
                "body": json.dumps({"error": "Google OAuth not configured"})
            }
        
        # Refresh token
        token_url = "https://oauth2.googleapis.com/token"
        token_data = {
            "refresh_token": refresh_token,
            "client_id": client_id,
            "client_secret": client_secret,
            "grant_type": "refresh_token"
        }
        
        response = requests.post(token_url, data=token_data, timeout=30)
        tokens = response.json()
        
        if "error" in tokens:
            return {
                "statusCode": 400,
                "headers": headers,
                "body": json.dumps({"error": tokens.get("error_description", "Token refresh failed")})
            }
        
        return {
            "statusCode": 200,
            "headers": headers,
            "body": json.dumps({
                "access_token": tokens.get("access_token"),
                "expires_in": tokens.get("expires_in")
            })
        }
        
    except Exception as e:
        return {
            "statusCode": 500,
            "headers": headers,
            "body": json.dumps({"error": str(e)})
        }
